File: strings/longest_duplicate_substring.rb
# @param {String} s
# @return {String}
def longest_dup_substring(s)
    l = 0
    r = 1
    res = ""

    while r < s.size
        substr = s[l...r]
        if s[(l + 1)..].include?(substr)
            res = substr if r - l > res.size
            r += 1
            next
        end

        l += 1
        r += 1 if l == r
    end

    res
end

File: strings/longest_duplicate_substring.py
class Solution:
    def longestDupSubstring(self, s: str) -> str:
        l = 0
        r = 1
        res = ""
        n = len(s)

        while r < n:
            if s[l:r] in s[l+1:]:
        
                if r - l > len(res):
                    res = s[l:r]
                r+=1
                continue
            l+=1
            if l == r:
                r+=1
        return res    